types/ramda/test/when-tests.ts

```diff
@@ -1,25 +1,33 @@
 import * as R from 'ramda';
 
 () => {
-  // $ExpectType (a: string) => string
-  const truncate = R.when(
-    (str: string) => str.length > 10,
-    (str: string) => str.slice(0, 10) + '…',
-  );
+    // $ExpectType <A extends string = string>(a: A) => string | A
+    const truncate = R.when(
+        (str: string) => str.length > 10,
+        (str: string) => str.slice(0, 10) + '…',
+    );
 
-  // $ExpectType string
-  truncate('12345'); // => '12345'
-  // $ExpectType string
-  truncate('0123456789ABC'); // => '0123456789…'
+    // $ExpectType string
+    truncate('12345'); // => '12345'
 
-  // $ExpectType (a: number | undefined) => number | undefined
-  const addOneIfNotNil = R.when<undefined | number, number>(
-      x => x != null,
-      R.add(1)
-  );
+    // $ExpectType string
+    truncate('0123456789ABC'); // => '0123456789…'
 
-  // $ExpectType number | undefined
-  const nil = addOneIfNotNil(undefined);
-  // $ExpectType number | undefined
-  const two = addOneIfNotNil(1);
+    // $ExpectType <A extends number | undefined = number>(a: A) => string | A
+    const testBroaderType = R.when(
+        (x: number | undefined) => x != null,
+        (x: number | string) => 'some text',
+    );
```

Comment:
The predicate function accepts `undefined`, yet the `whenTrueFn` doesn't accept `undefined` which doesn't seem right to me. The new typing now allows for invalid functions to be passed.

When you run this with the types that should be correct, there is actually nothing wrong with the old typing:

```typescript
whenOld(
        (x: number | undefined) => x != null,
        (x: number | string | undefined) => 'some text',
);
```

Reply:
Maybe you meant to support this?
```typescript
const testBroaderType = R.when(
        (x: number | undefined): x is number => x != null,
        (x: number | string) => 'some text',
    );
```

Reply:
That is the idea. `number` is in both so it is completely usable. Let's say you wanna reuse some functions, but their types do not extend but have overlap.
Function `testBroaderType` will have type `(number) => string | number` which is useful, since we can pass number in new function `testBroaderType`. With anonymous functions it may seem no use(bcs you can just add types to the union), but when you wanna reuse other functions there is sense to use such typings.

Reply:
But that's already working with the old typing, you can do either this:
```typescript
const testBroaderType = R.when(
        (x: number | undefined) => x != null,
        (x: number | string | undefined) => 'some text',
);
```
or this:
```typescript
const testBroaderType = R.when(
        (x: number) => x != null,
        (x: number | string) => 'some text',
);
```

They both work. It just wasn't working because of the undefined in the predicate, but not in the `whenTrueFn`

Reply:
Ok, just saw what you are trying to do. Give me a second to go through it again.

Reply:
I would say, since for the problem you are trying to solve you need to manually pass type parameters anyway, it would be easier to just to this (with the old typing):
```typescript
const testBroaderType = R.when<number, string>(
        (x: number | undefined) => x != null,
        (x: number | string) => 'some text',
);
```

Reply:
Yep we can solve it like that. Though again we need to provide it manually. And imagine for big types... in pipelines. We would need to construct it manually, manually perform an intersection, just a small example(but it can be a lot bigger type):
```ts
    const test3 = R.when(
        (x: { a: string; b: string, aa: {a: number}, bb: {e: string} }) => x != null,
        (x: { a: string; c: number, d: string }) => "some text"
    );
```

Reply:
Again, I do not think it is very very critical. But useful anyway. My main idea is this: https://github.com/ramda/ramda/issues/3257 to make it work with `strict` typescript environment.

Reply:
Do we have `discord` to talk? I think talking could be very useful. My discord: `Valerka#2430`

Reply:
I understand what you are trying to do. My issue is, that the new typing allows for all kind of wrong arguments to be passed, while making the return type much harder to read, just to avoid manual type arguments in an edge case situation.

```typescript
const testBroaderType = when(
        (x: number | undefined) => x != null,
        (x: number | string) => 'some text',
    );

testBroaderType(undefined) // <- This shouldn't be allowed, since the `whenTrueFn` can't 
                           //    handle `undefined`, but your current typing allows this!
```

Its cool if the function is easier to work with, but not at the cost of being less correct than before.

Reply:
```typescript
const testBroaderType = when(
        (x: number) => x != null,
        (x: number) => 1,
    );
```
Return type of this is: `<A extends number = number>(a: A) => number | A` which is less than ideal.

Reply:
Yep, that was my bad, I missed the case with undefined. Pushed fix.

Reply:
I made it compatible with the previous version.
It works the same, but for cases when previous version did not work(needed specify types manually), now types are inferred automatically. Also added few more test cases to show it.